includes/api/class-wc-rest-order-refunds-controller.php

```diff
@@ -0,0 +1,519 @@
+	 *
+	 * @param WC_Order_Refund $refund Comment object.
+	 * @return array Links for the given order refund.
+	 */
+	protected function prepare_links( $refund ) {
+		$order_id = $refund->post->post_parent;
+		$base     = str_replace( '(?P<order_id>[\d]+)', $order_id, $this->rest_base );
+		$links    = array(
+			'self' => array(
+				'href' => rest_url( sprintf( '/%s/%s/%d', $this->namespace, $base, $refund->id ) ),
+			),
+			'collection' => array(
+				'href' => rest_url( sprintf( '/%s/%s', $this->namespace, $base ) ),
+			),
+			'up' => array(
+				'href' => rest_url( sprintf( '/%s/orders/%d', $this->namespace, $order_id ) ),
+			),
+		);
+
+		return $links;
+	}
+
+	/**
+	 * Query args.
+	 *
+	 * @param array $args
+	 * @param WP_REST_Request $request
+	 * @return array
+	 */
+	public function query_args( $args, $request ) {
+		// Set post_status.
+		$args['post_status'] = 'any';
+
+		return $args;
+	}
+
+	/**
+	 * Create a single item.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_Error|WP_REST_Response
+	 */
+	public function create_item( $request ) {
+		if ( ! empty( $request['id'] ) ) {
+			return new WP_Error( "woocommerce_rest_{$this->post_type}_exists", sprintf( __( 'Cannot create existing %s.', 'woocommerce' ), $this->post_type ), array( 'status' => 400 ) );
+		}
+
+		$order_data = get_post( (int) $request['order_id'] );
+
+		if ( empty( $order_data ) ) {
+			return new WP_Error( 'woocommerce_rest_invalid_order', __( 'Order is invalid', 'woocommerce' ), 400 );
+		}
+
+		if ( 0 > $request['amount'] ) {
+			return new WP_Error( 'woocommerce_rest_invalid_order_refund', __( 'Refund amount must be greater than zero.', 'woocommerce' ), 400 );
+		}
+
+		$api_refund = is_bool( $request['api_refund'] ) ? $request['api_refund'] : true;
+
+		$data = array(
+			'order_id'   => $order_data->ID,
+			'amount'     => $request['amount'],
+			'line_items' => $request['line_items'],
+		);
+
+		// Create the refund.
+		$refund = wc_create_refund( $data );
+
+		if ( ! $refund ) {
+			return new WP_Error( 'woocommerce_rest_cannot_create_order_refund', __( 'Cannot create order refund, please try again.', 'woocommerce' ), 500 );
+		}
+
+		// Refund via API.
+		if ( $api_refund ) {
+			if ( WC()->payment_gateways() ) {
+				$payment_gateways = WC()->payment_gateways->payment_gateways();
+			}
+
+			$order = wc_get_order( $order_data );
+
+			if ( isset( $payment_gateways[ $order->payment_method ] ) && $payment_gateways[ $order->payment_method ]->supports( 'refunds' ) ) {
+				$result = $payment_gateways[ $order->payment_method ]->process_refund( $order_id, $refund->get_refund_amount(), $refund->get_refund_reason() );
+
+				if ( is_wp_error( $result ) ) {
+					return $result;
+				} elseif ( ! $result ) {
+					return new WP_Error( 'woocommerce_rest_create_order_refund_api_failed', __( 'An error occurred while attempting to create the refund using the payment gateway API', 'woocommerce' ), 500 );
```

Comment:
I think most of our other errors include punctuation so it would be good to be consistent/follow a style (ditto on line 276 "Order is invalid".